### PR TITLE
Add delegation functionality

### PR DESCRIPTION
**Pull request by soerenbf**
## Purpose

The delegation protocol for the node introduces new versions of: 

- `BlockSummary`
- `AccountInfo`

Along with adding new queries for: 

- Getting a list of baker IDs
- Getting the status of a specific pool.

The purpose of this PR is to have these reflected in the types and functions exposed by the SDK.

## Changes

- Updated types and added relevant functions to access the functionality described above.
- Bumped major version, as the type changes will be breaking (although the JS-only changes are additive, and could be contained in a minor version).

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: src/types.ts [hunk 1]
```diff
@@ -257,52 +264,144 @@ export interface FinalizationData {
...
+    max: N;
+}
+
+/**
+ * Used as index of a reward period, or as a number of reward periods.
+ * A reward period spans a number of epochs, as defined in chain parameters as "rewardPeriodLength"
+ */
+export type RewardPeriod = bigint;
+/** Index of an epoch, or number of epochs. */
+export type Epoch = bigint;
```
Comment (shjortConcordium) [inline src/types.ts:285]: Do we actually use these aliases to signify index?

Comment (soerenbf) [reply src/types.ts:285]: The comment is basically taken from the haskell implementation of the corresponding type

File: src/types.ts [hunk 2]
```diff
@@ -525,29 +710,171 @@ export interface InitialAccountCredential {
...
+}
+
+interface PoolStatusWrapper<T extends keyof typeof PoolStatusType, S> {
+    poolType: T;
+    poolStatus: S;
+}
+
+export interface BakerPoolStatusDetails {
+    bakerId: BakerId;
+    bakerAddress: Hex;
```
[the remark this answers is not in the document]
Comment (soerenbf) [reply src/types.ts:816]: I thought they were :D

File: src/types.ts [hunk 3]
```diff
@@ -567,7 +894,10 @@ export interface AccountInfo {
         Versioned<InitialAccountCredential | NormalAccountCredential>
     >;
 
+    // Only one of either accountBaker or accountDelegation can be active at any time.
```
[the remark this answers is not in the document]
Comment (soerenbf) [reply src/types.ts]: Done.

File: src/util.ts [hunk 4]
```diff
@@ -23,6 +23,10 @@ function intToString(jsonStruct: string, keys: string[]): string {
     return result;
 }
 
+export function intListToStringList(jsonStruct: string): string {
```
[the remark this answers is not in the document]
Comment (soerenbf) [reply src/util.ts:26]: Added some unit tests. Please check if you think they suffice.

File: src/client.ts [hunk 5]
```diff
@@ -600,6 +613,103 @@ export default class ConcordiumNodeClient {
...
+        blockHash: string,
+        bakerId: BakerId
+    ): Promise<BakerPoolStatus | undefined>;
+    /**
+     * Gets the status of either a baker, if a baker ID is supplied, or the L-pool if left undefined.
+     * @param blockHash the block hash the status at
+     * @param [bakerId] the ID of the baker to get the status for. If left undefined, the status of the L-pool is returned.
+     * @returns The status of the corresponding pool.
+     */
+    async getPoolStatus(
```
Comment (shjortConcordium) [inline src/client.ts:660]: Is this header necessary? :thinking:

File: test/client.test.ts [hunk 6]
```diff
@@ -409,7 +410,9 @@ test('account info with baker details, and with no pending change', async () =>
         throw new Error('Test failed to find account info');
     }
 
-    const bakerDetails = accountInfo.accountBaker;
+    const bakerDetails = isBakerAccount(accountInfo)
```
Comment (shjortConcordium) [inline test/client.test.ts:413]: Would it not make more sense to do:
```
 if (!isBakerAccount(accountInfo)) {
  throw error
 }
 const bakerDetails = accountInfo.accountBaker
```
maybe?

**Review by shjortConcordium — COMMENTED** [review]
LGTM, but lets test it on stagenet, when it is updated.

File: src/client.ts (hunk 5 above)
Comment (soerenbf) [reply src/client.ts:660]: When you look at the resulting types, only the function type definitions, and NOT the implementation, is present. I needed the dynamic version of the function in addition to the others.

File: src/types.ts [hunk 7]
```diff
@@ -357,6 +357,9 @@ export interface PoolParametersV1 {
 }
 
 export interface TimeParametersV1 {
+    /**
+     * In epochs
+     */
     rewardPeriodLength: Epoch;
```
Comment (shjortConcordium) [inline src/types.ts:363]: :thinking:

Comment (soerenbf) [reply src/types.ts]: Haha, the type aliases are removed in the outputted types, so this serves as documentation in it's place.

File: src/client.ts [hunk 8]
```diff
@@ -600,6 +616,140 @@ export default class ConcordiumNodeClient {
...
+            intToStringTransformer(bigInts)
+        );
+    }
+
+    /**
+     * Retrieve list of bakers on the network.
+     * @param blockHash the block hash to get the smart contact instances at
+     * @returns A JSON list of baker IDs
+     */
+    async getBakerList(blockHash: string): Promise<BakerId[] | undefined> {
```
Comment (shjortConcordium) [inline src/client.ts:661]: When I try this, I get a parse error,
it tries to convert "0,1,2,3,4" to a Bigint :thinking:

File: src/client.ts [hunk 9]
```diff
@@ -600,6 +616,140 @@ export default class ConcordiumNodeClient {
...
+     * @returns The status of the L-pool.
+     */
+    async getPoolStatus(blockHash: string): Promise<LPoolStatus | undefined>;
+    /**
+     * Gets the status a baker.
+     * @param blockHash the block hash the status at
+     * @param bakerId the ID of the baker to get the status for.
+     * @returns The status of the corresponding baker pool.
+     */
+    async getPoolStatus(
```
Comment (shjortConcordium) [inline src/client.ts:690]: If I don't provide a bakerId, I get a BakerPool type back, with bakerId 0.
When I do provide a bakerId, I get the LPool?

File: src/types.ts [hunk 10]
```diff
@@ -342,7 +470,7 @@ export interface UpdateQueue {
     queue: UpdateQueueQueue;
 }
 
-interface UpdateQueues {
+interface UpdateQueuesCommon {
```
Comment (shjortConcordium) [inline src/types.ts:473]: I guess this should not contain bakerStakeThreshold, as it only appears in V0

File: src/types.ts [hunk 11]
```diff
@@ -359,18 +487,94 @@ interface UpdateQueues {
...
-    updateQueues: UpdateQueues;
+/**
+ * Used from protocol version 1-3
+ */
+export type UpdateQueuesV0 = UpdateQueuesCommon;
+
+/**
+ * Used from protocol version 4
+ */
+export interface UpdateQueuesV1 extends UpdateQueuesCommon {
```
Comment (shjortConcordium) [inline src/types.ts:498]: Missing poolParameters

File: src/types.ts (hunk 10 above)
Comment (shjortConcordium) [reply src/types.ts:473]: I guess it is not the same as the authorizations

File: src/client.ts (hunk 9 above)
Comment (soerenbf) [reply src/client.ts:690]: Yes.. I experience this as well; might be an issue with the grpc function.

File: src/types.ts (hunk 10 above)
Comment (soerenbf) [reply src/types.ts:473]: Good catch!

Comment (soerenbf) [reply src/types.ts:473]: Had a talk with @td202 , and the authorizations prop is kept the same, while the update queue is renamed. Will add a comment to explain on Authorizations interface.

**Comment by soerenbf** [comment]
The added unit tests will fail, unless using a client targeting current stagenet version (at time of writing).

File: test/testHelpers.ts [hunk 12]
```diff
@@ -9,11 +9,11 @@ import { MobileWalletExport } from '../src/wallet/types';
  * Creates a client to communicate with a local concordium-node
  * used for automatic tests.
  */
-export function getNodeClient(): ConcordiumNodeClient {
+export function getNodeClient(address = '127.0.0.1'): ConcordiumNodeClient {
```
Comment (shjortConcordium) [inline test/testHelpers.ts:12]: You changed this to be a parameter, but is it ever used?

Comment (soerenbf) [reply test/testHelpers.ts:12]: Was to make it easier to target a specific node (e.g. stagenet) for some tests, without messing up the rest.

**Review by shjortConcordium — APPROVED** [review]

File: CHANGELOG.md [hunk 13]
```diff
@@ -1,5 +1,18 @@
 # Changelog
 
+## Unreleased
```
Comment (shjortConcordium) [inline CHANGELOG.md:3]:
```suggestion
## 0.7.0 2022-03-09
```

File: CHANGELOG.md [hunk 14]
```diff
@@ -1,5 +1,18 @@
...
+
+### Added
+
+- Support for getting baker list from node.
+- Support for getting status of Baker-/L-pool (required node to have protocol version 4 or later)
+- Helper functions for determining the version of `BlockSummary` and nested types. 
+
```
Comment (shjortConcordium) [inline CHANGELOG.md:10]:
```suggestion
- Support for initiate contract and update contract with parameters.

```
:upside_down_face:

**Pull request merged**